### DISCUSSION TITLE
Making members of an object available for use in a class (like Mixin)

### DISCUSSION THREAD

**Discussion by gulshan** [dotnet/csharplang#4772] [Language Ideas]
This proposal would be a C# syntactic sugar, much like the `using static` feature. The proposal is- a class can make members of any of its members (members of member) available for use, throughout the class, using some keyword.  The keyword can be `using` or `with` or any new/old keyword. I'm using `with` for demonstration-

Demo
---
Given the following class-
```csharp
class Adder
{
    public int x { get; init; }
    public int y { get; init; } 
    public int Add() => x + y;
}
```
This class-
```csharp
class ShowAdder
{
    private Adder adder = new Adder { x = 1, y = 1 };
    
    public void Show()
    {
        Console.WriteLine($"{adder.x} + {adder.y} = {adder.Add()}");
    }
}
```
can be transformed to this-
```csharp
class ShowAdder
{
    with public Adder adder = new Adder { x = 1, y = 1 };
    
    public void Show()
    {
        // No need to use `adder.` to access members of adder
        Console.WriteLine($"{x} + {y} = {Add()}");
    }
}
```
There would be no change is the actual class membership. There would be no effect outside the class. In this case, the consumers of the `ShowAdder` class will not see any change in its membership.

The __name resolution__ can follow the rules of `using static` (prioritize own member of same name). Or, it can be more restrictive and show error/warning on ambiguous names.

Advantages
---
- Provides Mixin-like functionality, without changing actual class membership.
- This will facilitate easier __composition over inheritance__ code.
- This can replace some inheritance and partial class codes, actually making the implementation more restrictive and safer.
- Some common C# scenarios like services/views in a thin-controllers" would become shorter and easier to follow.

Disadvantages
---
- One may not like the implicit behavior.
- Can be abused to to some extent. (add 10 "with" members?)

**Comment by HaloFour** [comment]
If all this does is "flatten" the type so that you don't need to qualify it I don't find it to be terribly compelling.  It adds a lot of complexity to how an identifier is resolved within the class, especially if the "mixin" members conflicts with instance members of the class or other "mixins", in which case you'd still be likely be forced to qualify the members with `this` or the name of the "mixin".

**Reply by gulshan** [reply]
>If all this does is "flatten" the type so that you don't need to qualify it

Exactly.
And regarding name resolution, as I've said, we can just follow the rules of `using static`.

**Comment by spydacarnage** [comment]
You say that you wouldn't want any consumer of the `ShowAdder` class to see the extra items from `Adder`.  So why make it `public`?

If that wasn't an issue, and you wanted those items to appear in the class, consider Source Generators...

**Reply by HaloFour** [reply]
Source generators sound like they would be an interesting way to enable this:

```cs
partial class ShowAdder
{
    [Mixin(Visibility = Visibility.Private)]
    private Adder adder = new Adder { x = 1, y = 1 };
    
    public void Show()
    {
        Console.WriteLine($"{x} + {y} = {adder.Add()}");
    }
}
```

would generate:

```cs
partial class ShowAdder
{
    private int x => adder.x;
    private int y => adder.y;
}
```

Which would allow the members to be accessed as instance members and without qualification.

**Reply by gulshan** [reply]
The proposal is, this would just be a syntactic sugar and would not change anything in the actual class membership. The `adder` in the `ShowAdder` could be private/public/anything. The point is, outside the class, it is all same.

As I don't want them as actual class members, just want them to use as other members within the class, partial classes and code generation is not what I am asking for. Rather this feature would replace some partial class usage. Like the partial classes used in winforms by designer. Then the full form can be generated and then put into another plain view class making the form object's members available to the view class.

**Reply by spydacarnage** [reply]
But partial classes are, at the end of the day, all compiled into one class, with all the members of all partials in one.

Ultimately, what this appears to boil down to, is you wanting a private member in your class and don't want to prefix calls to it with the member name.  Because, if it _was_ public, you'd be wanting to access `new ShowAdder().x`, even though the `ShowAdder` class doesn't have an `x` property...  That sounds like magic to me...

**Reply by gulshan** [reply]
> if it _was_ public, you'd be wanting to access `new ShowAdder().x`, even though the `ShowAdder` class doesn't have an `x` property... That sounds like magic to me...

Actually I do not want the ShowAdder class to have and member `x`, regardless of `adder` here being public or private. So `new ShowAdder().x` should not be possible. With this feature, a public `adder` would support `new ShowAdder().adder.x`, just like now, but not `new ShowAdder().x`. The scope of this feature is just within the class. Just like `using static` is scoped to a file.

**Reply by spydacarnage** [reply]
So, externally, you want to specify the member name, but _internally_, you want to ignore it.

So, this really is just a feature about not typing the member name...  I don't think that's going to gain _any_ traction, tbh.

**Reply by HaloFour** [reply]
Source generators can emit private members which would make them inaccessible outside of the class.  As those members are very tiny the JIT can certainly inline them, which would result in zero runtime impact.  This is a solution that you could have today that gives you the functionality that you desire.  It's my opinion that a language change to do what you requested would never be considered because qualifying "adder" is idiomatic and far from onerous.

**Comment by VardanKosakyan** [comment]
What if there are 2 child  objects of type Adder , both decorated with "with" keyword or some other type with x property ?

**Reply by gulshan** [reply]
Any ambiguous situation like this should generate a compilation error I guess?
